### DISCUSSION TITLE
Improve the documentation on CSRF protection

### DISCUSSION THREAD

**Discussion by lucabtz** [nextauthjs/next-auth#8919] [Help]
### What is the improvement or update you wish to see?

The `next-auth` docs are missing information about Cross Site Request Forgery and about how to accurately protect authenticated API routes against it.

### Is there any context that might help us understand?

I have been trying to understand whether CSRF protection comes already with `next-auth` or not. For context consider an API route as this
```typescript
export async function POST(request: NextRequest) {
  try {
    const session = await getServerSession(authOptions);

    if (!session || !session.user) {
      return new NextResponse(JSON.stringify({ error: "Unauthorized" }), {
        status: 401,
      });
    }

    // here goes the state-changing code ...

  } catch (error: any) {
    console.error(error);
    return new NextResponse(
      JSON.stringify({ error: "Internal server error" }),
      { status: 500 }
    );
  }
}
```

There is no documentation on what exactly `getServerSession` does regarding CSRF (e.g. look at [1,2]). I do not understand whether the only protection against CSRF here is the `SameSite=Lax` cookie attribute or more is going on. I read that `next-auth` uses the doubly submitted cookie pattern, but I submit no cookie in the request body and the API request is successful (I suppose the doubly submitted cookie is used in the `/api/auth/[...nextauth]` requests only).

Summing up what is not clear from the documentation to me is:
- is the above example code safe against CSRF on an older browser which does not support `SameSite=Lax`? 
- if the answer to the above question is yes: why is it secure? What protection mechanism is in place?
- if the answer to the first question is no: why is not secure? How can I implement a protection mechanism to make it secure?

**References**
[1]: https://next-auth.js.org/tutorials/securing-pages-and-api-routes
[2]: https://next-auth.js.org/configuration/nextjs

### Does the docs page already exist? Please link to it.

https://next-auth.js.org/tutorials/securing-pages-and-api-routes

**Comment by ThangHuuVu** [comment]
hi @ekardnam , does this page answer your questions? https://next-auth.js.org/getting-started/rest-api#get-apiauthcsrf. `next-auth` handles CSRF protection for `/api/auth/[...nextauth]` requests automatically. Usage of `getServerSession` inside other API routes is safe against CSRF since it's within the server. In this case, you need to implement your own CSRF for the client-server interaction through the API route.

**Reply by lucabtz** [reply]
> does this page answer your questions?

Not really no. I saw that page already, but it does not address the issue imho.

> Usage of getServerSession inside other API routes is safe against CSRF since it's within the server. In this case, you need to implement your own CSRF for the client-server interaction through the API route.

I'm not sure what you mean by this. Part of your answer suggests an API route like the one above is not safe against CSRF which is what I was supsecting. I think what I get is that `getServerSession` does no CSRF token check so it is not clear to me what you mean when you say it is safe against CSRF.

I think the page https://next-auth.js.org/tutorials/securing-pages-and-api-routes should mention this clearly since it is about securing API routes and it looks like the example is actually not fully secure. At least mention that CSRF protection has to be implemented and is not given automatically by NextAuth. It's plenty of resources which explain how to authenticate a request for a NextAuth route and none mentions CSRF, it is not a huge deal since the `SameSite=Lax` cookie policy is being used, however it'd be nice if the documentation mentioned this explicitly.

It would also be nice to know if there is a way to tell NextAuth to only check Bearer tokens when using JWTs. That would avoid dealing with CSRF tokens entirely.
